### PR TITLE
[Deterministic] Support TP-Invariant Inference

### PR DESCRIPTION
**Pull request by fopdoodle8**
## Motivation

In RL training, the rollout engine and the training engine often use different tensor parallel sizes. Rollout engines (e.g. SGLang) typically use larger TP (e.g. TP=8) to maximize throughput, while training engines such as FSDP use TP=1 for better memory efficiency. This TP size mismatch leads to noticeable numerical gaps in predicted probabilities between rollout and training, introducing a training–inference mismatch that can undermine on-policy RL and harm stability.

This PR integrates [Tree Based Invariant Kernels (TBIK)](https://tbik.super.site/) to ensure bit-wise deterministic results across different TP sizes, eliminating TP induced numerical discrepancies.

## Modifications

- **Tree Based Invariant Kernels (`tp_invariant_ops.py`)**
   - `matmul_tp_persistent` for TP Invariant MatMul which is registered as a pytorch op `torch.ops.tp_inv_ops.matmul_tp_inv`.
   - `tree_all_reduce_sum` for inter-GPU all reduce.
- **Communication Op (`communication_op.py`)**
   - Add `tensor_model_parallel_tree_all_reduce` for tree all reduce which uses `tree_all_reduce_sum` in `tp_invariant_ops.py`.
- **Row Parallel Layer (`linear.py`)**
   - Use `torch.ops.tp_inv_ops.matmul_tp_inv` for MatMul when tp invariant is enabled and apply `tensor_model_parallel_tree_all_reduce` for all reduce.
- **Layer Communicator (`communicator.py`)**
   - For some models that do not reduce results in `RowParallelLinear` but in  `_gather_hidden_states_and_residual`, we apply `tensor_model_parallel_tree_all_reduce` for all reduce when tp invariant is enabled.
- **Server Args (`server_args.py`)**
   - Add `fsdp_tp` into `rl_on_policy_target`. Use `rl-on-policy-target=fsdp_tp` to enable TP Invariant Mode.
- **Unittest Files (`/python/sglang/test/srt/test_deterministic_tp_util.py`, `/python/sglang/test/srt/test_deterministic_tp.py`, `/test/srt/deterministic_tp.py`)**
   - Add unittest files for tp invariant following `test_deterministic.py` 

## Accuracy Tests

**Use GSM8K fewshot**
```shell
python3 -m sglang.test.few_shot_gsm8k --num-questions 200
```
without TP Invariant Mode
```shell
python3 -m sglang.launch_server --model Qwen/Qwen3-8B --enable-deterministic-inference --tp-size 4
```

```
Accuracy: 0.950
Invalid: 0.000
Latency: 66.181 s
Output throughput: 366.920 token/s
```

with TP Invariant Mode
```shell
python3 -m sglang.launch_server --model Qwen/Qwen3-8B --enable-deterministic-inference --tp-size 4 --rl-on-policy-target fsdp_tp
```

```
Accuracy: 0.950
Invalid: 0.000
Latency: 57.249 s
Output throughput: 419.414 token/s
```

**TP Invariant UnitTest**

Host 1
```
python3 -m sglang.launch_server \
 --model "Qwen/Qwen3-8B" \
 --port 30000 \
 --attention-backend triton \
 --tensor-parallel-size 1 \
```

Host 2
```
python3 -m sglang.launch_server \
 --model "Qwen/Qwen3-8B" \
 --port 30001 \
 --attention-backend triton \
 --tensor-parallel-size 2 \
```

Test Prefix Mode
``` shell
python3 -m sglang.test.test_deterministic_tp --host-a localhost --port-a 30000 --host-b localhost --port-b 30001 --test-mode prefix --n-start 1 --n-trials 50 --return-logprob
```

without TP Invariant Mode
```
Comparing Host A (http://localhost:30000) vs Host B (http://localhost:30001)

--- Trial 1 (Batch Size 1) ---
Prompt Prefix 511: 1 samples
  ❌ Logprob Mismatch Sample 0: Logprob mismatch at position 0: -1.5945285558700562 vs -1.5941681861877441 (diff: 0.0003603696823120117)

--- Trial 2 (Batch Size 2) ---
Prompt Prefix 1: 2 samples
  ❌ Logprob Mismatch Sample 0: Logprob mismatch at position 0: -2.809950351715088 vs -2.7998223304748535 (diff: 0.010128021240234375)
  ❌ Logprob Mismatch Sample 1: Logprob mismatch at position 0: -2.809950351715088 vs -2.7998223304748535 (diff: 0.010128021240234375)
```

with TP Invariant Mode
```
Comparing Host A (http://localhost:30000) vs Host B (http://localhost:30001)

--- Trial 1 (Batch Size 1) ---
Prompt Prefix 511: 1 samples
  ✅ Logprobs match

--- Trial 2 (Batch Size 2) ---
Prompt Prefix 1: 2 samples
  ✅ Logprobs match
```

Test Radix Cache
```shell
python3 -m sglang.test.test_deterministic_tp --host-a localhost --port-a 30000 --host-b localhost --port-b 30001 --test-mode radix_cache
```

without TP Invariant Mode
```
--- Running Sequence on Host A (http://localhost:30000) ---

--- Running Sequence on Host B (http://localhost:30001) ---

============================================================
Comparing Host A vs Host B Results
============================================================
1. Initial Generation Comparison:
   ❌ Mismatch: Logprob mismatch at position 0: -2.677564859390259 vs -2.689919948577881 (diff: 0.01235508918762207)
2. Cached Generation Comparison:
   ❌ Mismatch: Logprob mismatch at position 0: -0.0054801469668745995 vs -0.005380434915423393 (diff: 9.971205145120621e-05)
3. Uncached Generation Comparison:
   ❌ Mismatch: Logprob mismatch at position 0: -0.0054801469668745995 vs -0.005380434915423393 (diff: 9.971205145120621e-05)

✗✗✗ Radix Cache Consistency Test Failed! ✗✗✗
```

with TP Invariant Mode
```
--- Running Sequence on Host A (http://localhost:30000) ---

--- Running Sequence on Host B (http://localhost:30001) ---

============================================================
Comparing Host A vs Host B Results
============================================================
1. Initial Generation Comparison:
   ✅ Match
2. Cached Generation Comparison:
   ✅ Match
3. Uncached Generation Comparison:
   ✅ Match

✓✓✓ Radix Cache Consistency Test Passed! ✓✓✓
```

Test Different Attention Backend with TP Invariant Mode

FA3
```
python3 -m unittest test_deterministic_tp.TestFa3Deterministic

Ran 2 tests in 151.136s

OK
```

Triton
```
python3 -m unittest test_deterministic_tp.TestTritonDeterministic

Ran 2 tests in 152.834s

OK
```

FlashInfer
```
python3 -m unittest test_deterministic_tp.TestFlashinferDeterministic

Ran 2 tests in 153.172s

OK
```

## Benchmarking and Profiling

without TP Invariant Mode
```shell
python3 -m sglang.bench_offline_throughput --model-path Qwen/Qwen3-8B --num-prompts 10 --tp-size 4 --enable-deterministic-inference
```
```
====== Offline Throughput Benchmark Result =======
Backend:                                 engine    
Successful requests:                     10        
Benchmark duration (s):                  11.10     
Total input tokens:                      1997      
Total generated tokens:                  2798      
Last generation throughput (tok/s):      58.03     
Request throughput (req/s):              0.90      
Input token throughput (tok/s):          179.89    
Output token throughput (tok/s):         252.05    
Total token throughput (tok/s):          431.94    
==================================================
```

with TP Invariant Mode
```shell
python3 -m sglang.bench_offline_throughput --model-path Qwen/Qwen3-8B --num-prompts 10 --tp-size 4 --enable-deterministic-inference --rl-on-policy-target fsdp_tp
```
```
====== Offline Throughput Benchmark Result =======
Backend:                                 engine    
Successful requests:                     10        
Benchmark duration (s):                  10.30     
Total input tokens:                      1997      
Total generated tokens:                  2798      
Last generation throughput (tok/s):      62.88     
Request throughput (req/s):              0.97      
Input token throughput (tok/s):          193.87    
Output token throughput (tok/s):         271.63    
Total token throughput (tok/s):          465.50    
==================================================
```

## Current Scope

1. TP Invariant Mode does not support DP Attention yet.
2. MoE models are not supported yet.
3. RowParallelLinear layers with `bias=True` and either `skip_all_reduce=True` or `reduce_results=False` are not supported yet.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [x] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

File: python/sglang/srt/tp_invariant_ops/tp_invariant_ops.py [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+
+                    carry_over = (table_value == count_value_added).to(tl.int1)
+
+                    if count_value_added > 1:
+                        if level == 0:
+                            acc = acc1 + acc
+                        elif level == 1:
+                            acc = acc2 + acc
+                        elif level == 2:
+                            acc = acc3 + acc
```
Comment (PopSoda2002) [inline python/sglang/srt/tp_invariant_ops/tp_invariant_ops.py]: Can you make `acc1` `acc2` more descriptive?

**Review by PopSoda2002 — COMMENTED** [review]
NIce work!

**Review by yizhang2077 — APPROVED** [review]
LGTM overall

File: python/sglang/srt/tp_invariant_ops/tp_invariant_ops.py [hunk 2]
```diff
@@ -0,0 +1,373 @@
+import contextlib
```
Comment (yizhang2077) [inline python/sglang/srt/tp_invariant_ops/tp_invariant_ops.py:3]: we could add reference for TBIK

File: python/sglang/srt/tp_invariant_ops/tp_invariant_ops.py (hunk 1 above)
Comment (fopdoodle8) [reply python/sglang/srt/tp_invariant_ops/tp_invariant_ops.py]: Renamed `acc1/acc2/acc3` in commit `9cf1726`.

File: python/sglang/srt/tp_invariant_ops/tp_invariant_ops.py (hunk 2 above)
Comment (fopdoodle8) [reply python/sglang/srt/tp_invariant_ops/tp_invariant_ops.py:3]: Added reference for TBIK in commit `9cf1726`.
